### DISCUSSION TITLE
[Question] QuadCopter training with camera

### DISCUSSION THREAD

**Discussion by JulienHansen** [isaac-sim/IsaacLab#1451] [Q&A] [answered]
Hi,

Apologies if this question has been asked before or if the answer is obvious. I’m still relatively new to reinforcement learning, robotics, and simulation, having started just a month ago.

I’m currently working on a reinforcement learning task for a drone that needs to navigate through a level to reach a desired position on the opposite side. To achieve this, I’ve added a camera to the drone for perception.

My question is: how should I approach training for such an application? In your opinion, what would be the fastest and most effective way to implement this?

Would it be better to:

1. Build my own library and custom wrappers?
2. Modify an existing library like RSL-RL?
3. Consider another approach entirely?

Thanks in advance for your responses

**Comment by shaoxiang** [comment]
I also really want to know that currently the quadcopter example only has a direct version, not a manage_based version, and I don't know if it's possible to implement a manage_based version because it doesn't seem to be possible to implement it through ActionsCfg's EffortActionCfg, etc. Therefore, I personally think that it may be difficult to implement a ready-made visual quadcopter.

**Comment by RandomOakForest** [comment]
This is a great discussion. Will move it into our _Discussions_ section for the team to follow upon. Thanks for posting this.

**Comment by Toni-SM** [comment]
@JulienHansen in order to provide an appropriate solution there is necessary to know in details what is the expected observation and action space you want to use

**Reply by JulienHansen** [reply]
@Toni-SM
I’ve made some modifications to the quadcopter environment provided by IsaacLab. Specifically, I’ve added a TiledCamera with rgb data_type to the drone. My task is to navigate through a level, and I want the drone to use the images it perceives to become aware of obstacles in the level and avoid them.

To achieve this in the least amount of time possible, I was thinking of modifing the actor_critic file in the rsl_rl library to include a CNN capable of processing these images. Additionally, I will have to adjust the wrapper to handle a new entry in the get_observation function and not just the "policy" entry

In my current setup, I plan to pass the drone’s intrinsic values and the desired position as a tensor under the key "policy" and the camera image under the key "cam_image" to PPO.

I hope i'm clear enough in my description regarding the action space it's the same as in the quadcopterEnv provided

**Reply by Toni-SM** [reply]
let's split the RL solution in 2 parts.
The first one is the environment, the second one the RL library.

For the environment, according to your description you need a composite observation space (for the camera image, robot data and target position, etc..). It could be a gymnasium `Tuple` (information stored positionally) or Dict (information stored by key).

Currently support for other spaces types rather than `Box` is only available in the Direct workflow. See [DirectRLEnvCfg.observation_space](https://isaac-sim.github.io/IsaacLab/main/source/api/lab/omni.isaac.lab.envs.html#omni.isaac.lab.envs.DirectRLEnvCfg.observation_space)

Is this okay? If so, could you please provide an initial space definition?

**Reply by JulienHansen** [reply]
For the moment i'm using a Dict, here is my get_observation function in my application : 

    def _get_observations(self) -> dict:
        cam_images = self.scene["camera"].data.output["rgb"][..., :3]
        desired_pos_b, _ = subtract_frame_transforms(
            self._robot.data.root_state_w[:, :3], self._robot.data.root_state_w[:, 3:7], self._desired_pos_w
        )

        state = torch.cat([
            self._robot.data.root_lin_vel_b,        
            self._robot.data.root_ang_vel_b,        
            self._robot.data.projected_gravity_b,   
            desired_pos_b,
        ], dim=-1)

        return {
            "policy": state,
            "Camera": cam_images
        }
        
If i remember correctly it is explain in the doc that this function needs to return the observations under the "policy" key, so to extend this function for my camera purpose i extended the dictionnary and i was thinking about modifying the rsl_rl wrapper 
and actor_critic file as explained above to gain time and not having to do to much code. 
I hope i'm answering your reply and not forgetting or misunderstanding your demand  has english is not my first language.

**Reply by Toni-SM** [reply]
In the Direct workflow the observation space value (fundamental or composite) needs, and it can, to be defined under the policy key. E.g.:

```python
observation = {
    "policy": {
        "robot-state": state,
        "camera": camera,
    }
}
```

Now, the second part, the RL library.
Currently, only **skrl** (version 1.4.0: `develop` branch, not released yet) support learning using any space (fundamental or composite). See https://skrl.readthedocs.io/en/develop/api/utils/model_instantiators.html#inputs

You can use it... or modify the rsl_rl or rl_games Isaac Lab wrappers and libraries source code to support it. 

Do you have any initial network structure? If  so, I can help you defining it in the skrl's agent config file.

**Reply by JulienHansen** [reply]
Ohhhhh okay i get it now i didn't know we could implement  composite observation space like that it's awesome ! 
Regarding the initial network structure except if you are referring to the skrl_config file like this one : 

https://github.com/isaac-sim/IsaacLab/blob/d603b67fd190eb3cc5c86d71b0a0aeadef5ad1be/source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/manager_based/locomotion/velocity/config/go1/agents/skrl_flat_ppo_cfg.yaml#L4

i don't have one and i could clearly benefit from you help. 🙂

**Reply by Toni-SM** [reply, marked as answer]
Well, I asked for an initial network structure to get an idea of a possible **skrl** configuration definition.
Again, if you are going to use **skrl** you need to use the version 1.4.0 (not released yet): [develop](https://github.com/Toni-SM/skrl/tree/develop) branch.

```bash
./isaaclab.sh -p -m pip install git+https://github.com/Toni-SM/skrl.git@develop
```

<hr>

Here is an example:

![image](https://github.com/user-attachments/assets/481cc14f-373b-4b61-b59b-1d3133ce4ba5)

* left network:

    ```yaml
    network:
      - name: features_extractor
        input: permute(STATES["camera"], (0, 3, 1, 2))  # PyTorch NHWC -> NCHW
        layers:
          - conv2d: {out_channels: 32, kernel_size: 8, stride: 4, padding: 0}
          - conv2d: {out_channels: 64, kernel_size: 4, stride: 2, padding: 0}
          - conv2d: {out_channels: 64, kernel_size: 3, stride: 1, padding: 0}
          - flatten
        activations: relu
      - name: net_state
        input: STATES["robot-state"]
        layers: [128, 128, 64]
        activations: elu
      - name: net
        input: concatenate([features_extractor, net_state])
        layers: [512, 256, 128]
        activations: elu
    output: ACTIONS
    ```

* right network:
    ```yaml
    network:
      - name: features_extractor
        input: permute(STATES["camera"], (0, 3, 1, 2))  # PyTorch NHWC -> NCHW
        layers:
          - conv2d: {out_channels: 32, kernel_size: 8, stride: 4, padding: 0}
          - conv2d: {out_channels: 64, kernel_size: 4, stride: 2, padding: 0}
          - conv2d: {out_channels: 64, kernel_size: 3, stride: 1, padding: 0}
          - flatten
        activations: relu
      - name: net
        input: concatenate([features_extractor, STATES["robot-state"]])
        layers: [512, 256, 128]
        activations: elu
    output: ACTIONS
    ```

Refer to https://skrl.readthedocs.io/en/develop/api/utils/model_instantiators.html#inputs for more details about model definitions in **skrl** using Isaac Lab .yaml configs

Also, see the skrl [skrl_camera_ppo_cfg.yaml](https://github.com/isaac-sim/IsaacLab/blob/main/source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/direct/cartpole/agents/skrl_camera_ppo_cfg.yaml) file for the `Isaac-Cartpole-RGB-Camera-Direct-v0` example (CNN)

**Reply by JulienHansen** [reply]
Thank you so much it much more clear now how i can trainmy drone !

**Reply by nianba** [reply]
> In the Direct workflow the observation space value (fundamental or composite) needs, and it can, to be defined under the policy key. E.g.:
> 
> ```python
> observation = {
>     "policy": {
>         "robot-state": state,
>         "camera": camera,
>     }
> }
> ```
> 
> Now, the second part, the RL library. Currently, only **skrl** (version 1.4.0: `develop` branch, not released yet) support learning using any space (fundamental or composite). See https://skrl.readthedocs.io/en/develop/api/utils/model_instantiators.html#inputs
> 
> You can use it... or modify the rsl_rl or rl_games Isaac Lab wrappers and libraries source code to support it.
> 
> Do you have any initial network structure? If so, I can help you defining it in the skrl's agent config file.

Is there any similar setting in manage based env?

**Comment by Sanjay1911** [comment]
Hi @Toni-SM 

Thanks for mentioning few keypoints for training with Vision based RL in Isaaclab. 
I've another question as my observation space has a point cloud data from Isaaclab Raycasting sensor. My idea is to use a PointNet Style encoder for feature extraction and feeding it to observation space. But from the [Instantiators](https://skrl.readthedocs.io/en/develop/api/utils/model_instantiators.html#inputs) page, it looks like I donot have many options to use the custom abstraction heads of PointNet++.

Do you've any suggestions or ideas regarding this problem statement.

**Reply by ducmngx** [reply]
How do you include LiDAR in your RL training? I'm trying to have 2D laser scan for Go2 navigation and I haven't figured out how

**Reply by Sanjay1911** [reply]
You can check the [Height Scan observation](https://github.com/isaac-sim/IsaacLab/blob/1e273226879c341f99b56a0c14e3a9c05c98766a/source/isaaclab/isaaclab/envs/mdp/observations.py#L234) for reference. For my case, since my field of view is narrow and to learn an embedding, I first downsample outer points and pass it via a MLP (PointNet style). But of course it depends on your observation dimension and your scan area.

**Reply by ducmngx** [reply]
Yes, but from what I have seen, the height scan is an Ray Caster with prim_path_mesh as "World/GroundPlane". I am not sure how I can scan the environment instead of the ground.

**Reply by Sanjay1911** [reply]
Ahh now I get what you mean. I locally merged #1886 into my current workspace. This PR adds dynamic and multi-mesh support and saved a whole lot of time for me. Give it a go!

**Reply by DavoodSZ** [reply]
Hi @Sanjay1911 

I also want to have a custom network to process and encode the observation before feeding it to the RL policy. Were you able to implement that?

Thanks

**Reply by bhenriquezsoto** [reply]
Hey @Sanjay1911 were you able to solve it / get any response for this issue??

Thanks in advance!

**Reply by Sanjay1911** [reply]
Hi @DavoodSZ1993 ,

Apologies for my late reply. It actually depends on the object/asset whose information you want to encode for the observations. 
In my case since it's a vessel mesh, the network couldn't learn anything from a simple pointnet based architecture. 

But if you want to encode camera images, there are few examples in Cartpole Showcase environments under Direct RL tasks

**Reply by Sanjay1911** [reply]
Hi @bhenriquezsoto,
Are you encoding sensor data / image data?
Since I want my agent to learn dynamic collision avoidance, I use a discrete bin strategy to identify collision spaces. Encoding was an overkill for my goal.

**Reply by bhenriquezsoto** [reply]
> Hi @bhenriquezsoto,
> 
> Are you encoding sensor data / image data?
> 
> Since I want my agent to learn dynamic collision avoidance, I use a discrete bin strategy to identify collision spaces. Encoding was an overkill for my goal. 

I'm trying to do active perception using a LiDAR connected to a drone through a gimbal (initially moving only the LiDAR with a pre established routine for the drone and then moving both according to a policy), I would like to train this (probably) partial RL policy in IsaacLab but I haven't seen much cases of success in this domain.

Do you happen to have any resources for this other than the one you recommended already?

**Reply by Sanjay1911** [reply]
> > Hi @bhenriquezsoto,
> > Are you encoding sensor data / image data?
> > Since I want my agent to learn dynamic collision avoidance, I use a discrete bin strategy to identify collision spaces. Encoding was an overkill for my goal.
> 
> I'm trying to do active perception using a LiDAR connected to a drone through a gimbal (initially moving only the LiDAR with a pre established routine for the drone and then moving both according to a policy), I would like to train this (probably) partial RL policy in IsaacLab but I haven't seen much cases of success in this domain.
> 
> Do you happen to have any resources for this other than the one you recommended already?

Unfortunately not in this domain. But I know this amazing work from @kousheekc - https://github.com/kousheekc/isaac_drone_racer 
My strategy is  partly based on this paper - https://arxiv.org/abs/2503.00496

**Reply by bhenriquezsoto** [reply]
@Sanjay1911 perfect, I will look into both resources, thank you so much, have a great day!
